### PR TITLE
Allow for custom markets in binance

### LINKED ISSUE #1073: Find way to speed up binance historical trade queries
## Problem Definition

Binance historical queries need you to iterate over every single possible market that binance offers:
https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#account-trade-list-user_data

This is slow.

There seems to be no endpoint that offers a way to query all pairs in one go.

## Task

Find a way to improve the speed of this query.

### PR DESCRIPTION
**Pull request by yabirgb**
Closes #1073

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -1031,6 +1031,7 @@ Setup or remove an exchange
    :reqjson string api_secret: The api secret with which to setup the exchange
    :reqjson string passphrase: An optional passphrase, only for exchanges, like coinbase pro, which need a passphrase.
    :reqjson string kraken_account_type: An optional setting for kraken. The type of the user's kraken account. Valid values are "starter", "intermediate" and "pro".
+   :reqjson list binance_markets: An optional setting for binance and binanceus. A list of string for markets that should be queried.
```
Comment (LefterisJP) [inline docs/api.rst:1034]: Can you add an example market here?

File: docs/api.rst [hunk 2]
```diff
@@ -8106,3 +8107,61 @@ ERC20 token info
...
+      }
+
+User selected Binance markets
+================================
+
+.. http:get:: /api/(version)/exchanges/binance/pairs/(exchange account name)
+
+   Doing a GET to this endpoint will return the market pairs that the user has selected to be queried at binance.
+
+   .. note::
```
Comment (LefterisJP) [inline docs/api.rst:8147]: I think this note is not needed here. Having an exchange account name as a requirement implies that.

File: docs/changelog.rst [hunk 3]
```diff
@@ -2,6 +2,8 @@
 Changelog
 =========
 
+* :feature:`1073` Binance users will now be able to select which markets should be queried.
```
Comment (LefterisJP) [inline docs/changelog.rst:5]:
```suggestion
* :feature:`1073` Binance users will now be able to select which markets should be queried for trades history considerably improving the speed of binance history queries.
```

File: rotkehlchen/constants/misc.py [hunk 4]
```diff
@@ -24,6 +24,8 @@
 KRAKEN_API_VERSION = '0'
 BINANCE_BASE_URL = 'binance.com/'
 BINANCEUS_BASE_URL = 'binance.us/'
+BINANCE_SETTING_SEPARATOR = ' '
```
Comment (LefterisJP) [inline rotkehlchen/constants/misc.py:27]: move these constants either inside binance only or in a constants file in a binance subdir if they are only used there

File: rotkehlchen/db/dbhandler.py [hunk 5]
```diff
@@ -2186,6 +2187,7 @@ def add_exchange(
             api_secret: ApiSecret,
             passphrase: Optional[str] = None,
             kraken_account_type: Optional[KrakenAccountType] = None,
+            binance_markets: Optional[List[str]] = None,  # pylint: disable=W0613
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2190]: Ehmm ... why are you ignoring the unused argument from pylint? Shouldn't you actually use the `binance_markets` argument in this function to save it in the DB?

File: rotkehlchen/db/dbhandler.py [hunk 6]
```diff
@@ -2267,6 +2270,15 @@ def edit_exchange(
                 self.conn.rollback()
                 raise InputError(f'Could not update DB user_credentials_mappings due to {str(e)}') from e  # noqa: E501
 
+        location_is_binance = location in (Location.BINANCE, Location.BINANCEUS)
+        if location_is_binance and binance_markets is not None:
+            try:
+                account_name = new_name if new_name is not None else name
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2276]: Replace "account_name". It's a confusing name for this variable. This is the exchange name.

```suggestion
                exchange_name = new_name if new_name is not None else name
```

File: rotkehlchen/db/dbhandler.py [hunk 7]
```diff
@@ -2338,6 +2350,101 @@ def get_exchange_credentials_extras(self, name: str, location: Location) -> Dict
 
         return extras
 
+    def add_binance_pairs(self, name: str, pairs: List[str], location: Location) -> None:
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2353]: Why is this function not used anywhere and why is only `set_binance_pairs()` used?

File: rotkehlchen/db/dbhandler.py [hunk 8]
```diff
@@ -2338,6 +2350,101 @@ def get_exchange_credentials_extras(self, name: str, location: Location) -> Dict
...
+                    name,
+                    location.serialize_for_db(),
+                    BINANCE_MARKETS_KEY,
+                    str(data),
+                ),
+            )
+        self.conn.commit()
+        self.update_last_write()
+
+    def set_binance_pairs(self, name: str, pairs: List[str], location: Location) -> None:
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2390]: I don't know which of the 2 functions you want to use but why not just use json? Just stringify the json list and deserialize the string to json when reading. json dumps and loads

it would simplify all these functions you wrote here

File: rotkehlchen/exchanges/binance.py [hunk 9]
```diff
@@ -794,7 +794,7 @@ def query_balances(self) -> ExchangeQueryBalances:
             self.first_connection()
             returned_balances: DefaultDict[Asset, Balance] = defaultdict(Balance)
             returned_balances = self._query_spot_balances(returned_balances)
-            if self.name != str(Location.BINANCEUS):
+            if self.location != Location.BINANCEUS:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/binance.py:797]: no need to do this, this is already done in bugfixes and may even conflict during merging bugfixes to develop, so please don't

File: rotkehlchen/exchanges/binance.py [hunk 10]
```diff
@@ -1075,3 +1077,13 @@ def query_online_margin_history(
...
+
+    def get_selected_pairs(self) -> List[BinancePair]:
+        pairs = self.db.get_binance_pairs(self.name, self.location)
+        validated_pairs = []
+        if pairs:
+            for pair in pairs:
+                known_pair = self._symbols_to_pair.get(pair)
+                if known_pair is not None:
+                    validated_pairs.append(known_pair)
+        return validated_pairs
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/binance.py:1089]: This is too complicated. If you want to follow this approach make a set of the `_symbols_to_pair` and a set of the `pairs` and then get their union. Faster and easier to read.

But ... shouldn't you actually verify that a pair is valid **before** saving it in the database?

Which I don't see you doing at all.

On the other hand even if you verify it they may delist the pair in the meantime, but I would like to verify any data going in the DB.

**Review by LefterisJP — COMMENTED** [review]
Some important comments and things that need changing.

Also why did you commit a new tests/data/global/global.d?

Also for frontend I guess @kelsos would need to review.

I did try it. And I get console errors:

```
vue.runtime.esm.js:619 [Vue warn]: Property or method "label" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.

found in

---> <BinancePairsSelector> at src/components/helper/BinancePairsSelector.vue
       <VForm>
         <ExchangeKeysForm> at src/components/settings/api-keys/ExchangeKeysForm.vue
           <VCard>
             <VThemeProvider>
               <VBottomSheet>
                 <BigDialog> at src/components/dialogs/BigDialog.vue
                   <ExchangeSettings> at src/components/settings/api-keys/ExchangeSettings.vue
                     <VTabItem>
                       <VTabsItems>
                         <VTabs>
                           <TabNavigation> at src/components/helper/TabNavigation.vue
                             <ApiKeys> at src/views/settings/ApiKeys.vue
                               <VMain>
                                 <VApp>
                                   <App> at src/App.vue
                                     <Root>
```

It would be much better if the user could see the assets they are trying to see the market for. So if it would be possible to split market into assets. But ... I do understand this is probaly more complicated for both backend and frontend and could be another issue in the future

File: frontend/app/src/components/helper/BinancePairsSelector.vue [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        :disabled="loading"
+        hide-details
+        hide-selected
+        hide-no-data
+        return-object
+        chips
+        clearable
+        :outlined="outlined"
+        :open-on-clear="false"
+        :label="label ? label : $t('blockchain_account_selector.default_label')"
```
Comment (kelsos) [inline frontend/app/src/components/helper/BinancePairsSelector.vue:19]: Shouldn't the label be replaced with something like
"Trade markets"? (extracted to the en.json)

File: frontend/app/src/components/helper/BinancePairsSelector.vue [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    const query = queryText.toLocaleLowerCase();
+    const pair = item.toLocaleLowerCase();
+
+    return pair.indexOf(query) > -1;
+  }
+}
+</script>
+
+<style scoped lang="scss">
+.blockchain-account-selector {
```
Comment (kelsos) [inline frontend/app/src/components/helper/BinancePairsSelector.vue:111]: Since this is a new components it would be better to also update the classes to reflect the component's name

File: frontend/app/src/components/helper/BinancePairsSelector.vue [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+            filter
+            close
+            @click:close="data.parent.selectItem(data.item)"
+          >
+            {{ data.item }}
+          </v-chip>
+        </template>
+        <template #item="data">
+          <div
+            class="blockchain-account-selector__list__item d-flex justify-space-between flex-grow-1"
```
Comment (kelsos) [inline frontend/app/src/components/helper/BinancePairsSelector.vue:42]: After the classes update the references to these classes would have to be updated too.

File: frontend/app/src/components/helper/BinancePairsSelector.vue [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  selection: string[] = [];
+  allMakerts: string[] = [];
+  multiple: boolean = true;
+
+  async mounted() {
+    this.queriedMarkets = await this.$api.queryBinanceUserMarkets(
+      this.name,
+      this.location
+    );
+    this.allMakerts = await this.$api.queryBinanceMarkets();
```
Comment (kelsos) [inline frontend/app/src/components/helper/BinancePairsSelector.vue:90]: Both these promises should be awaited in a try/catch, otherwise, any network error would blow.

For the loading that @LefterisJP mentioned you could add a `loading: boolean = false` property to the component and then change it to true/false when starting/ending this method, then you can pass that to the autocomplete as `:loading='loading'`

File: frontend/app/src/services/rotkehlchen-api.ts [hunk 15]
```diff
@@ -872,6 +872,22 @@ export class RotkehlchenApi {
...
 
+  queryBinanceMarkets(): Promise<string[]> {
+    return this.axios
+      .get<ActionResult<string[]>>('/exchanges/binance/pairs')
+      .then(handleResponse);
+  }
+
+  queryBinanceUserMarkets(name: string, location: string): Promise<string[]> {
+    return this.axios
+      .get<ActionResult<string[]>>('/exchanges/binance/pairs/' + name, {
```
Comment (kelsos) [inline frontend/app/src/services/rotkehlchen-api.ts:883]: this is a nitpick but would it be easy to change that to a template format instead? 

https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

File: rotkehlchen/db/dbhandler.py (hunk 7 above)
Comment (yabirgb) [reply rotkehlchen/db/dbhandler.py:2353]: Nice questions. This function only appends to the list of markets. For how we receive this information in the API it makes sense to use the `set_...` function. First I added this one and then when I had to call the API saw that the one needed was the other one. 

I didn't delete this one because thought that we may want at some point to autodetect markets and then this may be useful.

File: rotkehlchen/exchanges/binance.py (hunk 10 above)
Comment (yabirgb) [reply rotkehlchen/exchanges/binance.py:1089]: The problem of verifying this at the moment of inserting is that it implies a new query to the binance api and then I would need to make the request async in the frontend what is a bit more complicated.

**Comment by kelsos** [comment]
Frontend code looks fine to me 👍🏼

File: docs/api.rst [hunk 16]
```diff
@@ -8106,3 +8107,58 @@ ERC20 token info
    :statuscode 200: No critical error found.
    :statuscode 409: There is an error with the address.
    :statuscode 500: Internal rotki error.
+
+All Binance markets
+======================
+
+.. http:get:: /api/(version)/exchanges/binance/pairs
+
+   Doing a GET to this endpoint will return the market pairs available at binance.
```
Comment (LefterisJP) [inline docs/api.rst:8116]:
```suggestion
   Doing a GET to this endpoint will return all the market pairs available at binance.
```

File: rotkehlchen/api/v1/resources.py [hunk 17]
```diff
@@ -1576,3 +1581,17 @@ class ERC20TokenInfo(BaseResource):
...
+    def get(self) -> Response:
+        return self.rest_api.get_all_binance_pairs()
+
+
+class BinanceUserMarkets(BaseResource):
+
+    get_schema = BinanceMarketsUserSchema()
+
+    @use_kwargs(get_schema, location='json_and_query_and_view_args')
+    def get(self, name: str, location: Location) -> Response:
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:1596]: Why do you need the name in this schema? Binance will have same pairs for all api keys. Binance US also.

File: rotkehlchen/db/dbhandler.py [hunk 18]
```diff
@@ -2334,6 +2350,35 @@ def get_exchange_credentials_extras(self, name: str, location: Location) -> Dict
...
+        data = json.dumps(pairs)
+        cursor.execute(
+            'INSERT OR REPLACE INTO user_credentials_mappings '
+            '(credential_name, credential_location, setting_name, setting_value) '
+            'VALUES (?, ?, ?, ?)',
+            (
+                name,
+                location.serialize_for_db(),
+                BINANCE_MARKETS_KEY,
+                str(data),
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2364]: nitpick: Why do you need `str()` here. Json dumps should always return a string.

File: rotkehlchen/exchanges/manager.py [hunk 19]
```diff
@@ -274,3 +277,18 @@ def initialize_exchanges(
...
                 )
                 self.connected_exchanges[location].append(exchange_obj)
+
+    def get_all_binance_pairs(self) -> List[str]:
+        if Location.BINANCEUS in self.connected_exchanges:
+            binance = self.connected_exchanges[Location.BINANCEUS][0]
+            return binance._symbols_to_pair.keys()  # type: ignore
+        if Location.BINANCE in self.connected_exchanges:
+            binance = self.connected_exchanges[Location.BINANCE][0]
+            return binance._symbols_to_pair.keys()  # type: ignore
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/manager.py:287]: Prioritize normal binance

```suggestion
        if Location.BINANCE in self.connected_exchanges:
            binance = self.connected_exchanges[Location.BINANCE][0]
            return binance._symbols_to_pair.keys()  # type: ignore
        if Location.BINANCEUS in self.connected_exchanges:
            binance = self.connected_exchanges[Location.BINANCEUS][0]
            return binance._symbols_to_pair.keys()  # type: ignore
```

File: rotkehlchen/tests/api/test_exchanges.py [hunk 20]
```diff
@@ -1015,3 +1015,19 @@ def test_edit_exchange_credentials(rotkehlchen_api_server_with_exchanges):
...
+    response = requests.get(
+        api_url_for(
+            server,
+            'binanceavailablemarkets',
+        ),
+    )
+    result = assert_proper_response_with_result(response)
+    some_pairs = ['ETHUSDC', 'BTCUSDC', 'BNBBTC', 'FTTBNB']
+    for pair in some_pairs:
+        assert pair in result
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_exchanges.py:1033]: nitpick: Use set comparisons :)

File: rotkehlchen/tests/exchanges/test_binance.py [hunk 21]
```diff
@@ -769,3 +771,41 @@ def mock_response(url, timeout):  # pylint: disable=unused-argument
...
+
+    def counter():
+        count = 0
+        while True:
+            yield count
+            count += 1
+
+    count = counter()
+    p = re.compile(r'symbol=[A-Z]*')
+    seen = set()
```
Comment (LefterisJP) [inline rotkehlchen/tests/exchanges/test_binance.py:798]: You can do this much simpler by using `nonlocal`. Example here:

https://github.com/rotki/rotki/blob/8256eb4fd72feb12fee09bfc360feb07b6ce8bc0/rotkehlchen/tests/unit/test_inquirer.py#L63-L77

**Review by LefterisJP — COMMENTED** [review]
Some smaller comments.

File: rotkehlchen/api/v1/resources.py (hunk 17 above)
Comment (yabirgb) [reply rotkehlchen/api/v1/resources.py:1596]: Because the setting is set for pairs `(name, location)` and I find it nice since you can have multiple binance accounts and select different markets

File: rotkehlchen/tests/exchanges/test_binance.py (hunk 21 above)
Comment (yabirgb) [reply rotkehlchen/tests/exchanges/test_binance.py:798]: Thanks! I was wondering how to improve this test :+1:

**Review by LefterisJP — COMMENTED** [review]
So it looks fine to me but there is a frontend problem that needs fixing.

It's really annoying to have to delete the extra text before saving.

Can't you somehow make this extra text go away each time the user selects a pair? Or something along those lines?

**Comment by yabirgb** [comment]
Let me try again if I can figure it out

**Review by LefterisJP — APPROVED** [review]
LGTM. Well done! :clap:
